Source/Initialization/WarpXInitData.cpp

```diff
@@ -1236,3 +1255,135 @@ void WarpX::InitializeEBGridData (int lev)
     amrex::ignore_unused(lev);
 #endif
 }
+
+void
+WarpX::ReadExternalFieldsFromFile (std::string read_fields_from_path, MultiFab* mf,
+std::string F_name, std::string F_component)
+{
+    //! add_external: 0 for no exteranl field loading; 1/2/3 for E/B/both.
+    if (E_ext_grid_s=="read_from_file" && B_ext_grid_s=="read_from_file") {
+        add_external_fields = 3;
+    } else {
+        if (E_ext_grid_s=="read_from_file") { add_external_fields = 1; }
+        if (B_ext_grid_s=="read_from_file") { add_external_fields = 2; }
+    }
+
+    // Get WarpX domain info
+    auto& warpx = WarpX::GetInstance();
+    amrex::Geometry const& geom = warpx.Geom(0);
+    const amrex::RealBox& real_box = geom.ProbDomain();
+    const auto dx = geom.CellSizeArray();
+    amrex::IntVect nodal_flag = mf->ixType().toIntVect();
+
+    // Loop over boxes
+    for (MFIter mfi(*mf, TilingIfNotGPU()); mfi.isValid(); ++mfi)
+    {
+
+        // Read external field openPMD data
+        auto series = openPMD::Series(read_fields_from_path, openPMD::Access::READ_ONLY);
+        auto iseries = series.iterations.begin()->second;
+        auto F = iseries.meshes[F_name];
+        auto offset = F.gridGlobalOffset();
+        auto d = F.gridSpacing<long double>();
+        auto FC = F[F_component];
+        auto extent = FC.getExtent();
+
+        auto box = mfi.growntilebox();
+        auto lo = lbound(box);
+        auto hi = ubound(box);
+        const amrex::RealBox& real_box = geom.ProbDomain();
+        const auto dx = geom.CellSizeArray();
+
+        // Determine the chunk data that will be loaded.
+        // Now, the full range of data is loaded.
+        // Loading chunk data can speed up the process.
+        // Thus, `chunk_offset` and `chunk_extent` should be modified accordingly in another PR.
+#if defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+        openPMD::Offset chunk_offset = {0,0};
+        openPMD::Extent chunk_extent = {extent[0],extent[1]};
+#else
+        openPMD::Offset chunk_offset = {0,0,0};
+        openPMD::Extent chunk_extent = {extent[0],extent[1],extent[2]};
+#endif
+
+        auto FC_chunk_data = FC.loadChunk<double>(chunk_offset,chunk_extent);
+        series.flush();
+        auto FC_data = FC_chunk_data.get();
```

Comment:
Allocate a new multifab on pinned memory on CPU, such that GPU can also use it.
Copy FC_data to that multifab.
Any ideas on how to do it here?
@ax3l @RemiLehe

Reply:
@RemiLehe @ax3l This PR should be ready to review, and the GPU feature and relevant tests may be left as a follow-up PR.